Web Inspector: expanded sections of Storage should not collapse

the-chenergy opened the pull request:
#### e03b5777dbef981543833275d15e5bd4b3eaddd4
<pre>
Web Inspector: expanded sections of Storage should not collapse
<a href="https://rdar.apple.com/107687975">rdar://107687975</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=255071">https://bugs.webkit.org/show_bug.cgi?id=255071</a>

Reviewed by Devin Rousso.

Use WI.Setting to remember the expanded state of each FolderTreeElement
in the StorageSidebarPanel.

* Source/WebInspectorUI/UserInterface/Views/FolderTreeElement.js:
(WI.FolderTreeElement):
  - Optionally create and load a WI.Setting item for this element.

(WI.FolderTreeElement.prototype.collapse):
(WI.FolderTreeElement.prototype.expand):
  - Update the WI.Setting item on change.

* Source/WebInspectorUI/UserInterface/Views/StorageSidebarPanel.js:
(WI.StorageSidebarPanel.prototype._addDOMStorageObject):
(WI.StorageSidebarPanel.prototype._addDatabase):
(WI.StorageSidebarPanel.prototype._addIndexedDatabase):
(WI.StorageSidebarPanel.prototype._addCookieStorageObject):
(WI.StorageSidebarPanel.prototype._addFrameManifest):
(WI.StorageSidebarPanel.prototype._addStorageChild):
  - Give each of these folder elements an ID so that they create
    and use WI.Setting items.

Canonical link: <a href="https://commits.webkit.org/275783@main">https://commits.webkit.org/275783@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd99a975af96b4daa9ea8544736949896b2d31cf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42565 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21585 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44967 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45173 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38687 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25225 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18950 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35248 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43139 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18672 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36645 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16187 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16311 "Found 9 new test failures: imported/w3c/web-platform-tests/css/css-content/quotes-007.html, imported/w3c/web-platform-tests/css/css-transforms/transform3d-preserve3d-001.html, imported/w3c/web-platform-tests/css/css-view-transitions/hit-test-unrelated-element.html, imported/w3c/web-platform-tests/css/css-view-transitions/iframe-new-has-scrollbar.html, imported/w3c/web-platform-tests/css/css-view-transitions/iframe-old-has-scrollbar.html, imported/w3c/web-platform-tests/css/css-view-transitions/transition-in-empty-iframe.html, imported/w3c/web-platform-tests/html/semantics/forms/the-input-element/radio-disconnected-group-owner.html, imported/w3c/web-platform-tests/svg/painting/reftests/marker-path-012.svg, imported/w3c/web-platform-tests/xhr/setrequestheader-case-insensitive.htm (failure)") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37698 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/622 "Built successfully") | 
| | [❌ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38820 "Found 1 new API test failure: TestWebKitAPI.DataDetectorTests.LoadWKWebViewWithDataDetectorTypePhoneNumber (failure)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38027 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46656 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17384 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14330 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/41933 "Found 1 new test failure: webrtc/captureCanvas-webrtc-software-h264-high.html (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19003 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40546 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9551 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19067 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18648 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/FolderTreeElement.js:
```diff
@@ -25,11 +25,34 @@
         super(classNames, title, subtitle, representedObject, {hasChildren: true});
+
+        if (id) {
+            const settingName = "folder-tree-element-" + id;
```
Please use `let`. Web Inspector uses `const` only for truly [constant things](https://trac.webkit.org/wiki/WebInspectorCodingStyleGuide#Tokensspacingindentationsyntax).

rcaliman-apple commented on Source/WebInspectorUI/UserInterface/Views/StorageSidebarPanel.js:
```diff
@@ -340,7 +340,7 @@ WI.StorageSidebarPanel = class StorageSidebarPanel extends WI.NavigationSidebarP
             this.contentTreeOutline.removeChild(previousOnlyChild);
 
-            var folderElement = new WI.FolderTreeElement(folderName);
+            var folderElement = new WI.FolderTreeElement(folderName, null, folderId);
```
Help others quickly figure out what `null` is. This is a use case for `const` ;)
Also, please use `let` instead of `var` ([coding style guide](https://trac.webkit.org/wiki/WebInspectorCodingStyleGuide#Tokensspacingindentationsyntax)).

```suggestion
            const representedObject = null;
            let folderElement = new WI.FolderTreeElement(folderName, representedObject, folderId);
```

rcaliman-apple reviewed (commented):
Really nice 👍 A lot of people's work will be made easer by this. 
@karlcow will love it! :)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FolderTreeElement.js:
```diff
@@ -25,11 +25,36 @@
 WI.FolderTreeElement = class FolderTreeElement extends WI.GeneralTreeElement
 {
-    constructor(title, representedObject)
+    constructor(title, representedObject, id = null)
```
Style: we prefer using `options = {}` options "bags" instead of parameters with default values as it's _much_ easier to extend in the future
```
constructor(title, representedObject, options = {})
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FolderTreeElement.js:
```diff
@@ -25,11 +25,36 @@
+        if (id) {
+            let settingName = "folder-tree-element-" + id;
+            const defaultValue = false;
+            this._expandedSetting = new WI.Setting(settingName, defaultValue);
```
NIT: i know our normal style is to use `const` for things like this, but in this case it's pretty obvious that the second argument to `WI.Setting` is the default value so we normally just inline it
```js
if (options.id) {
    this._expandedSetting = new WI.Setting("folder-tree-element-" + options.id, false);
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FolderTreeElement.js:
```diff
@@ -25,11 +25,36 @@
+                this.expand();
+        }
+    }
+
+    collapse()
```
Style: there should be a `// Public` comment here to help identify these as public member functions
```js
}

// Public

collapse()
{
```

dcrousso reviewed (approved):
neat!

i coulda sworn we had this functionality somewhere, and it looks like it's in `WI.FolderizedTreeElement`.  we probably should have that use this instead
```js
folder = new WI.FolderTreeElement(settings.displayName, settings.representedObject, {id: settings.type});
```
(as well as removing `_folderTreeElementExpandedStateChange` and `_folderExpandedSettingMap`)

we may also want to add `{id: "..."}` to more places that use `WI.FolderTreeElement` (e.g. the "Results" folder in the Audit Tab, "Saved Recordings" in the Graphics Tab, most/all of the folders in the Sources Tab, etc.), though we should be careful to ensure that we dont create too many `WI.Setting` as otherwise we could unnecessarily "pollute" `localStorage` (e.g. if we saved the expanded state of each folder in the path to a resource in the Sources Tab we could end up with a lot of very specific `WI.Setting` keys)

if you'd rather do these two as followups then please file bugs for them so we have some sort of record :)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FolderTreeElement.js:
```diff
@@ -25,11 +25,36 @@
+        super.collapse();
+
+        if (this._expandedSetting)
+            this._expandedSetting.value = false;
```
NIT: just in case something happens in `super.collapse()` that would change the collapsed state, i'd do this instead
```
this._expandedSetting.value = this.expanded;
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/FolderTreeElement.js:
```diff
@@ -25,11 +25,36 @@
+        super.expand();
+
+        if (this._expandedSetting)
+            this._expandedSetting.value = true;
```
ditto

the-chenergy committed:
Web Inspector: expanded sections of Storage should not collapse
rdar://107687975
https://bugs.webkit.org/show_bug.cgi?id=255071

Reviewed by Devin Rousso.

Use WI.Setting to remember the expanded state of each FolderTreeElement
in the StorageSidebarPanel.

* Source/WebInspectorUI/UserInterface/Views/FolderTreeElement.js:
(WI.FolderTreeElement):
  - Optionally create and load a WI.Setting item for this element.

(WI.FolderTreeElement.prototype.collapse):
(WI.FolderTreeElement.prototype.expand):
  - Update the WI.Setting item on change.

* Source/WebInspectorUI/UserInterface/Views/StorageSidebarPanel.js:
(WI.StorageSidebarPanel.prototype._addDOMStorageObject):
(WI.StorageSidebarPanel.prototype._addDatabase):
(WI.StorageSidebarPanel.prototype._addIndexedDatabase):
(WI.StorageSidebarPanel.prototype._addCookieStorageObject):
(WI.StorageSidebarPanel.prototype._addFrameManifest):
(WI.StorageSidebarPanel.prototype._addStorageChild):
  - Give each of these folder elements an ID so that they create
    and use WI.Setting items.

Canonical link: https://commits.webkit.org/275783@main
